# Experiment 9e1add2185b302759ad4cf0c095b8fc84c5832ba4293203c8aa11f75c1316697
# Base 09528c39a8284d40e07ffa68732a642c71025cdf
diff --git a/include/utmarray.h b/include/utmarray.h
index 2db30cbf6..44c4d545b 100644
--- a/include/utmarray.h
+++ b/include/utmarray.h
@@ -30,11 +30,11 @@ public:
         this->maximumSizeValue = 0;
     }
 
-    T& operator[](int index)
+    T& operator[](int v)
     {
-        if (index > this->maximumSizeValue - 1)
-            this->resize(index + 1);
-        return this->value[index];
+        if (v > this->maximumSizeValue - 1)
+            this->resize(v + 1);
+        return this->value[v];
     }
 
     int enlarge(int amount)
@@ -51,7 +51,7 @@ public:
         return this->desiredNumberValue;
     }
 
-    int contains(const T& item);
+    int contains(const T& r);
 
     int position(T& item)
     {
@@ -63,29 +63,28 @@ public:
         return -1;
     }
 
-    int add(T item);
-
-    int remove(T item)
+    int add(T r)
     {
-        int i = 0;
-        int maximum = this->maximumSizeValue;
-
-        for (; i < maximum; i++) {
-            if (this->value[i] == item)
-                break;
-        }
-
-        if (i >= maximum)
+        if (this->contains(r) == 1)
             return 0;
+        if (this->numberValue > this->maximumSizeValue - 1)
+            this->resize(this->numberValue + 1);
+        this->value[this->numberValue] = r;
+        this->numberValue++;
+        return 1;
+    }
 
-        maximum--;
-        if (i < maximum) {
-            do {
-                this->value[i] = this->value[i + 1];
-                i++;
-            } while (i < this->maximumSizeValue - 1);
-        }
-
+    int remove(T r)
+    {
+        int i;
+        for (i = 0;
+             i < this->maximumSizeValue && this->value[i] != r;
+             i++)
+            ;
+        if (i >= this->maximumSizeValue)
+            return 0;
+        for (; i < this->maximumSizeValue - 1; i++)
+            this->value[i] = this->value[i + 1];
         this->numberValue--;
         if (this->numberValue < 0)
             this->numberValue = 0;
@@ -134,29 +133,26 @@ public:
         }
     }
 
-public:
-    // PDB marks resize protected. Reconstructed callers still require public
-    // access, so keep the body here while the source-access boundary is open.
-    int resize(int new_size)
+protected:
+    int resize(int s)
     {
-        T* new_value = new T[new_size];
+        T* new_value = new T[s];
         if (new_value == 0)
             return 0;
 
         int i;
         for (i = 0;
-             i < this->maximumSizeValue && i < new_size;
+             i < this->maximumSizeValue && i < s;
              i++) {
             new_value[i] = this->value[i];
         }
 
         delete[] this->value;
         this->value = new_value;
-        this->maximumSizeValue = new_size;
+        this->maximumSizeValue = s;
         return 1;
     }
 
-public:
     T* value;
     int numberValue;
     int desiredNumberValue;
@@ -164,52 +160,16 @@ public:
 };
 
 template <class T>
-int ManagedArray<T>::contains(const T& item)
+int ManagedArray<T>::contains(const T& r)
 {
     int i;
-    T wanted;
-    T current;
-
     for (i = 0; i < this->numberValue; i++) {
         if (i >= this->maximumSizeValue)
             return 0;
-        current = this->value[i];
-        wanted = item;
-        if (current == wanted)
+        if (this->value[i] == r)
             return 1;
     }
     return 0;
 }
 
-template <class T>
-int ManagedArray<T>::add(T val)
-{
-    int i;
-
-    for (i = 0; i < this->numberValue; i++) {
-        if (i >= this->maximumSizeValue)
-            break;
-        if (this->value[i] == val)
-            return 0;
-    }
-    if (this->numberValue > this->maximumSizeValue - 1) {
-        int newSize = this->numberValue + 1;
-        T* newValue = (T*)::operator new(newSize * sizeof(T));
-        if (newValue != 0) {
-            int j;
-            for (j = 0; j < this->maximumSizeValue; j++) {
-                if (j >= newSize)
-                    break;
-                newValue[j] = this->value[j];
-            }
-            ::operator delete(this->value);
-            this->value = newValue;
-            this->maximumSizeValue = newSize;
-        }
-    }
-    this->value[this->numberValue] = val;
-    this->numberValue++;
-    return 1;
-}
-
 #endif
diff --git a/src/taiinfmd.cpp b/src/taiinfmd.cpp
index cd749f29e..ff0438e97 100644
--- a/src/taiinfmd.cpp
+++ b/src/taiinfmd.cpp
@@ -465,16 +465,16 @@ TribeInformationAIModule::~TribeInformationAIModule()
                               this->maxResources[i] * sizeof(ResourceMemory));
             }
 
-            outFile = this->importantMiscs.numberValue;
+            outFile = this->importantMiscs.number();
             rge_write(file, &outFile, 4);
-            for (i = 0; i < this->importantMiscs.numberValue; i++) {
+            for (i = 0; i < this->importantMiscs.number(); i++) {
                 outFile = this->importantMiscs[i];
                 rge_write(file, &outFile, 4);
             }
 
-            outFile = this->itemsToDefend.numberValue;
+            outFile = this->itemsToDefend.number();
             rge_write(file, &outFile, 4);
-            for (i = 0; i < this->itemsToDefend.numberValue; i++) {
+            for (i = 0; i < this->itemsToDefend.number(); i++) {
                 outFile = this->itemsToDefend[i];
                 rge_write(file, &outFile, 4);
             }
@@ -662,39 +662,39 @@ int TribeInformationAIModule::save(int handle)
     for (int c = 0; c < this->maxAttackMemories; c++)
         rge_write(saved_handle, (char*)this->attackMemories + c * 0x18, 0x18);
 
-    handle = this->importantObjects.numberValue;
+    handle = this->importantObjects.number();
     rge_write(saved_handle, &handle, 4);
-    for (int d = 0; d < this->importantObjects.numberValue; d++) {
+    for (int d = 0; d < this->importantObjects.number(); d++) {
         handle = this->importantObjects[d];
         rge_write(saved_handle, &handle, 4);
     }
-    handle = this->importantUnits.numberValue;
+    handle = this->importantUnits.number();
     rge_write(saved_handle, &handle, 4);
-    for (int e = 0; e < this->importantUnits.numberValue; e++) {
+    for (int e = 0; e < this->importantUnits.number(); e++) {
         handle = this->importantUnits[e];
         rge_write(saved_handle, &handle, 4);
     }
-    handle = this->importantMiscs.numberValue;
+    handle = this->importantMiscs.number();
     rge_write(saved_handle, &handle, 4);
-    for (int f = 0; f < this->importantMiscs.numberValue; f++) {
+    for (int f = 0; f < this->importantMiscs.number(); f++) {
         handle = this->importantMiscs[f];
         rge_write(saved_handle, &handle, 4);
     }
-    handle = this->itemsToDefend.numberValue;
+    handle = this->itemsToDefend.number();
     rge_write(saved_handle, &handle, 4);
-    for (int g = 0; g < this->itemsToDefend.numberValue; g++) {
+    for (int g = 0; g < this->itemsToDefend.number(); g++) {
         handle = this->itemsToDefend[g];
         rge_write(saved_handle, &handle, 4);
     }
-    handle = this->playerBuildings.numberValue;
+    handle = this->playerBuildings.number();
     rge_write(saved_handle, &handle, 4);
-    for (int h = 0; h < this->playerBuildings.numberValue; h++) {
+    for (int h = 0; h < this->playerBuildings.number(); h++) {
         handle = this->playerBuildings[h];
         rge_write(saved_handle, &handle, 4);
     }
-    handle = this->playerObjects.numberValue;
+    handle = this->playerObjects.number();
     rge_write(saved_handle, &handle, 4);
-    for (int w = 0; w < this->playerObjects.numberValue; w++) {
+    for (int w = 0; w < this->playerObjects.number(); w++) {
         handle = this->playerObjects[w];
         rge_write(saved_handle, &handle, 4);
     }
@@ -1044,12 +1044,7 @@ int TribeInformationAIModule::gameIDOfResourceObject(int resType, int gathererId
                 if (obj->object_state != 2) goto next;
                 if (this->md->tacticalAI.numberGatherersWithTarget(obj->id) > 0) goto next;
                 objectId = obj->id;
-                k = 0;
-                while (k < this->farmsTaskedThisUpdate.numberValue) {
-                    if (k >= this->farmsTaskedThisUpdate.maximumSizeValue) break;
-                    if (this->farmsTaskedThisUpdate.value[k] == objectId) goto next;
-                    ++k;
-                }
+                if (this->farmsTaskedThisUpdate.contains(objectId) == 1) goto next;
             }
             if (resType == 1 && this->md->tacticalAI.numberGatherersWithTarget(obj->id) > 0) goto next;
             if ((resType == 3 || resType == 2) && this->md->tacticalAI.numberGatherersWithTarget(obj->id) > 3) goto next;
@@ -1088,7 +1083,7 @@ int TribeInformationAIModule::gameIDsOfObjectType(int objectType, int** ids, int
     *owners = 0;
     int maxCount = 10;
     int i = 0;
-    if (this->importantObjects.numberValue > 0) {
+    if (this->importantObjects.number() > 0) {
         do {
             RGE_Static_Object* obj = this->md->aiPlayer->world->object(
                 this->importantObjects[i]);
@@ -1122,7 +1117,7 @@ int TribeInformationAIModule::gameIDsOfObjectType(int objectType, int** ids, int
             }
 next:
             i++;
-        } while (i < this->importantObjects.numberValue);
+        } while (i < this->importantObjects.number());
     }
     return count;
 }
@@ -1320,7 +1315,7 @@ int TribeInformationAIModule::numberDocksInZone(int param_1, RGE_Zone_Map* param
     int count = 0;
     int i = 0;
 
-    if (this->playerBuildings.numberValue > 0) {
+    if (this->playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* obj = this->md->object(this->playerBuildings[i]);
             if (obj != 0 && obj->master_obj->id == 0x2D) {
@@ -1332,7 +1327,7 @@ int TribeInformationAIModule::numberDocksInZone(int param_1, RGE_Zone_Map* param
             }
 
             i++;
-        } while (i < this->playerBuildings.numberValue);
+        } while (i < this->playerBuildings.number());
     }
 
     return count;
@@ -1349,7 +1344,7 @@ int TribeInformationAIModule::closestDockDistance(int x, int y)
     i = 0;
     distance = -1;
 
-    if (this->playerBuildings.numberValue > 0) {
+    if (this->playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* obj = this->md->object(this->playerBuildings[i]);
             if (obj != 0 && obj->master_obj->id == 0x2D) {
@@ -1365,7 +1360,7 @@ int TribeInformationAIModule::closestDockDistance(int x, int y)
             }
 
             i++;
-        } while (i < this->playerBuildings.numberValue);
+        } while (i < this->playerBuildings.number());
     }
 
     return distance;
@@ -1732,82 +1727,19 @@ int TribeInformationAIModule::addImportantObject(RGE_Static_Object* obj, int rec
 
             marked = checkDefend(obj);
             id = obj->id;
-            i = 0;
-            for (i = 0; i < this->importantObjects.numberValue; ++i) {
-                if (i >= this->importantObjects.maximumSizeValue) break;
-                if (this->importantObjects.value[i] == id) goto important_unit;
-            }
-            if (this->importantObjects.numberValue > this->importantObjects.maximumSizeValue - 1) {
-                int newMaximum = this->importantObjects.numberValue + 1;
-                int* newValue = (int*)operator new(newMaximum * 4);
-                if (newValue != 0) {
-                    int j = 0;
-                    while (j < this->importantObjects.maximumSizeValue && j < newMaximum) {
-                        newValue[j] = this->importantObjects.value[j];
-                        ++j;
-                    }
-                    operator delete(this->importantObjects.value);
-                    this->importantObjects.value = newValue;
-                    this->importantObjects.maximumSizeValue = newMaximum;
-                }
-            }
-            this->importantObjects.value[this->importantObjects.numberValue] = id;
-            this->importantObjects.numberValue = this->importantObjects.numberValue + 1;
-
-important_unit:
+            this->importantObjects.add(id);
             if (importantUnit(oType) == 1) {
                 id = obj->id;
-                i = 0;
-                for (i = 0; i < this->importantUnits.numberValue; ++i) {
-                    if (i >= this->importantUnits.maximumSizeValue) break;
-                    if (this->importantUnits.value[i] == id) goto end;
-                }
-                if (this->importantUnits.numberValue > this->importantUnits.maximumSizeValue - 1) {
-                    int newMaximum = this->importantUnits.numberValue + 1;
-                    int* newValue = (int*)operator new(newMaximum * 4);
-                    if (newValue != 0) {
-                        int j = 0;
-                        while (j < this->importantUnits.maximumSizeValue && j < newMaximum) {
-                            newValue[j] = this->importantUnits.value[j];
-                            ++j;
-                        }
-                        operator delete(this->importantUnits.value);
-                        this->importantUnits.value = newValue;
-                        this->importantUnits.maximumSizeValue = newMaximum;
-                    }
-                }
-                this->importantUnits.value[this->importantUnits.numberValue] = id;
-                this->importantUnits.numberValue = this->importantUnits.numberValue + 1;
+                this->importantUnits.add(id);
             } else {
                 if (importantResource(oType) == 1) {
                     addResourceObject(obj);
                 } else if (importantMisc(oType) == 1) {
                     id = obj->id;
-                    i = 0;
-                    for (i = 0; i < this->importantMiscs.numberValue; ++i) {
-                        if (i >= this->importantMiscs.maximumSizeValue) break;
-                        if (this->importantMiscs.value[i] == id) goto end;
-                    }
-                    if (this->importantMiscs.numberValue > this->importantMiscs.maximumSizeValue - 1) {
-                        int newMaximum = this->importantMiscs.numberValue + 1;
-                        int* newValue = (int*)operator new(newMaximum * 4);
-                        if (newValue != 0) {
-                            int j = 0;
-                            while (j < this->importantMiscs.maximumSizeValue && j < newMaximum) {
-                                newValue[j] = this->importantMiscs.value[j];
-                                ++j;
-                            }
-                            operator delete(this->importantMiscs.value);
-                            this->importantMiscs.value = newValue;
-                            this->importantMiscs.maximumSizeValue = newMaximum;
-                        }
-                    }
-                    this->importantMiscs.value[this->importantMiscs.numberValue] = id;
-                    this->importantMiscs.numberValue = this->importantMiscs.numberValue + 1;
+                    this->importantMiscs.add(id);
                 }
             }
 
-end:
             if (importantUnit(oType) == 1 || marked == 1) {
                 addObjectMemory(obj);
             }
@@ -2437,55 +2369,11 @@ void TribeInformationAIModule::addObject(RGE_Static_Object* obj)
         }
 
         int id = object->id;
-        int count = this->playerObjects.numberValue;
-        int i = 0;
-        while (i < count && i < this->playerObjects.maximumSizeValue) {
-            if (this->playerObjects.value[i] == id) goto after_objects;
-            ++i;
-        }
-        if (count > this->playerObjects.maximumSizeValue - 1) {
-            int newMax = count + 1;
-            int* newArray = (int*)operator new(newMax * 4);
-            if (newArray != 0) {
-                i = 0;
-                while (i < this->playerObjects.maximumSizeValue && i < newMax) {
-                    newArray[i] = this->playerObjects.value[i];
-                    ++i;
-                }
-                operator delete(this->playerObjects.value);
-                this->playerObjects.value = newArray;
-                this->playerObjects.maximumSizeValue = newMax;
-            }
-        }
-        this->playerObjects.value[this->playerObjects.numberValue] = id;
-        ++this->playerObjects.numberValue;
-after_objects:
+        this->playerObjects.add(id);
 
         if (this->isBuilding(object) == 1) {
             id = object->id;
-            count = this->playerBuildings.numberValue;
-            i = 0;
-            while (i < count && i < this->playerBuildings.maximumSizeValue) {
-                if (this->playerBuildings.value[i] == id) goto after_buildings;
-                ++i;
-            }
-            if (count > this->playerBuildings.maximumSizeValue - 1) {
-                int newMax = count + 1;
-                int* newArray = (int*)operator new(newMax * 4);
-                if (newArray != 0) {
-                    i = 0;
-                    while (i < this->playerBuildings.maximumSizeValue && i < newMax) {
-                        newArray[i] = this->playerBuildings.value[i];
-                        ++i;
-                    }
-                    operator delete(this->playerBuildings.value);
-                    this->playerBuildings.value = newArray;
-                    this->playerBuildings.maximumSizeValue = newMax;
-                }
-            }
-            this->playerBuildings.value[this->playerBuildings.numberValue] = id;
-            ++this->playerBuildings.numberValue;
-after_buildings:
+            this->playerBuildings.add(id);
 
             short type = object->master_obj->id;
             if (type == 0x6d || type == 0x67 || type == 0x44) {
@@ -2500,48 +2388,8 @@ after_buildings:
 
 void TribeInformationAIModule::removeObject(int param_1)
 {
-    int idx;
-    int* ptr;
-
-    idx = 0;
-    while (idx < playerBuildings.maximumSizeValue) {
-        if (playerBuildings.value[idx] == param_1)
-            break;
-        idx++;
-    }
-
-    if (idx < playerBuildings.maximumSizeValue) {
-        if (idx < playerBuildings.maximumSizeValue - 1) {
-            do {
-                ptr = playerBuildings.value;
-                idx++;
-                ptr[idx - 1] = ptr[idx];
-            } while (idx < playerBuildings.maximumSizeValue - 1);
-        }
-        playerBuildings.numberValue--;
-        if (playerBuildings.numberValue < 0)
-            playerBuildings.numberValue = 0;
-    }
-
-    idx = 0;
-    while (idx < playerObjects.maximumSizeValue) {
-        if (playerObjects.value[idx] == param_1)
-            break;
-        idx++;
-    }
-
-    if (idx < playerObjects.maximumSizeValue) {
-        if (idx < playerObjects.maximumSizeValue - 1) {
-            do {
-                ptr = playerObjects.value;
-                idx++;
-                ptr[idx - 1] = ptr[idx];
-            } while (idx < playerObjects.maximumSizeValue - 1);
-        }
-        playerObjects.numberValue--;
-        if (playerObjects.numberValue < 0)
-            playerObjects.numberValue = 0;
-    }
+    playerBuildings.remove(param_1);
+    playerObjects.remove(param_1);
 
     RGE_Static_Object* obj = md->object(param_1);
     if (obj != 0) {
@@ -2974,7 +2822,7 @@ ObjectMemory* TribeInformationAIModule::objectToAttackByGroup(
                     wonder->id, commanderObject->weaponRange(), 0, 1, owner, 0x1B,
                     wallObjects) == 1) {
                 group->numberObjectsToDestroyValue = 0;
-                for (int i = 0; i < wallObjects.numberValue; ++i) {
+                for (int i = 0; i < wallObjects.number(); ++i) {
                     if (this->md->object(wallObjects[i]) != 0) {
                         this->addImportantObject(wallObjects[i]);
                         group->addObjectToDestroy(wallObjects[i]);
@@ -3167,7 +3015,7 @@ ObjectMemory* TribeInformationAIModule::objectToAttackByGroup(
         }
 
         group->numberObjectsToDestroyValue = 0;
-        for (int i = 0; i < wallObjects.numberValue; ++i) {
+        for (int i = 0; i < wallObjects.number(); ++i) {
             if (this->md->object(wallObjects[i]) != 0) {
                 this->addImportantObject(wallObjects[i]);
                 group->addObjectToDestroy(wallObjects[i]);
@@ -3403,48 +3251,31 @@ ObjectMemory* TribeInformationAIModule::higherPriorityObjectToDefend(int objId,
     *(unsigned char*)&objId = mainObj->currentZone();
     best = 0;
     group = this->defendPriority(type, group);
-    int i = 0;
-    if (this->maxImportantObjectMemory > 0) {
-        do {
-            ObjectMemory* entry = &this->importantObjectMemory[i];
-            int id = entry->id;
-            if (id != -1) {
-                int j = 0;
-                int defendCount = this->itemsToDefend.numberValue;
-                if (defendCount > 0) {
-                    while (j < defendCount) {
-                        if (j >= this->itemsToDefend.maximumSizeValue)
-                            break;
-                        if (this->itemsToDefend.value[j] == id) {
-                            if (entry->attackAttempts <= 0) {
-                                RGE_Static_Object* target = this->md->object(id);
-                                if (target != 0) {
-                                    if (*(unsigned char*)&objId != target->currentZone()) {
-                                        XYPoint pt;
-                                        pt.x = (int)target->world_x;
-                                        pt.y = (int)target->world_y;
-                                        if (mainObj->withinRangeOfZoneAtPoint(
-                                                *(unsigned char*)&objId,
-                                                mainObj->getSpeed(), pt) != 1)
-                                            break;
-                                    }
-                                    int pri = this->defendPriority(
-                                        this->importantObjectMemory[i].type,
-                                        this->importantObjectMemory[i].group);
-                                    if (group > pri) {
-                                        group = pri;
-                                        best = &this->importantObjectMemory[i];
-                                    }
-                                }
-                            }
-                            break;
-                        }
-                        j++;
-                    }
+    for (int i = 0; i < this->maxImportantObjectMemory; i++) {
+        ObjectMemory* entry = &this->importantObjectMemory[i];
+        int id = entry->id;
+        if (id != -1 && this->itemsToDefend.contains(id) == 1 &&
+            entry->attackAttempts <= 0) {
+            RGE_Static_Object* target = this->md->object(id);
+            if (target != 0) {
+                if (*(unsigned char*)&objId != target->currentZone()) {
+                    XYPoint pt;
+                    pt.x = (int)target->world_x;
+                    pt.y = (int)target->world_y;
+                    if (mainObj->withinRangeOfZoneAtPoint(
+                            *(unsigned char*)&objId,
+                            mainObj->getSpeed(), pt) != 1)
+                        continue;
+                }
+                int pri = this->defendPriority(
+                    this->importantObjectMemory[i].type,
+                    this->importantObjectMemory[i].group);
+                if (group > pri) {
+                    group = pri;
+                    best = &this->importantObjectMemory[i];
                 }
             }
-            i++;
-        } while (i < this->maxImportantObjectMemory);
+        }
     }
     return best;
 }
@@ -3540,11 +3371,7 @@ int TribeInformationAIModule::checkDefend(RGE_Static_Object* obj)
             if (result8 <= 0)
                 goto checkDefend_fail;
             tmp = obj->id;
-            ManagedArray<int>* list8 = &this->itemsToDefend;
-            if (list8->contains(tmp) == 1)
-                return 1;
-            (*list8)[list8->numberValue] = tmp;
-            this->itemsToDefend.numberValue++;
+            this->itemsToDefend.add(tmp);
             return 1;
         }
         case 32: {
@@ -3552,10 +3379,7 @@ int TribeInformationAIModule::checkDefend(RGE_Static_Object* obj)
             if (result32 <= 0)
                 goto checkDefend_fail;
             tmp = obj->id;
-            if (this->itemsToDefend.contains(tmp) == 1)
-                return 1;
-            this->itemsToDefend[this->itemsToDefend.numberValue] = tmp;
-            this->itemsToDefend.numberValue++;
+            this->itemsToDefend.add(tmp);
             return 1;
         }
         default:
@@ -3568,7 +3392,7 @@ checkDefend_fail:
 }
 
 int TribeInformationAIModule::numberItemsToDefend() {
-    return this->itemsToDefend.numberValue;
+    return this->itemsToDefend.number();
 }
 
 ObjectMemory* TribeInformationAIModule::objectToTradeWith(int param_1)
@@ -3826,7 +3650,7 @@ void TribeInformationAIModule::setupInfluenceMap(int mode, int objectType,
     if (mode == 1 || mode == 2 || mode == 3 || mode == 4 || mode == 0 ||
         mode == 5 || mode == 6 || mode == 8) {
         a = 0;
-        if (this->playerBuildings.numberValue > 0) {
+        if (this->playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* object = this->md->object(this->playerBuildings[a]);
             if (object != 0) {
@@ -4019,7 +3843,7 @@ void TribeInformationAIModule::setupInfluenceMap(int mode, int objectType,
             }
         building_next:
             a++;
-        } while (a < this->playerBuildings.numberValue);
+        } while (a < this->playerBuildings.number());
         }
     }
 
@@ -5630,7 +5454,7 @@ float TribeInformationAIModule::findClosestDropsite(RGE_Static_Object* obj, int
     dropsiteID = -1;
 
     i = 0;
-    if (this->playerBuildings.numberValue > 0) {
+    if (this->playerBuildings.number() > 0) {
         do {
             dropsite = this->md->object(this->playerBuildings[i]);
             if (dropsite != 0) {
@@ -5651,7 +5475,7 @@ float TribeInformationAIModule::findClosestDropsite(RGE_Static_Object* obj, int
             }
 
             i++;
-        } while (i < this->playerBuildings.numberValue);
+        } while (i < this->playerBuildings.number());
     }
 
     return best;
@@ -5667,7 +5491,7 @@ RGE_Static_Object* TribeInformationAIModule::findClosestReturnDropsite(RGE_Stati
     unsigned char unitZone = unit->currentZone();
     int i = 0;
 
-    if (this->playerBuildings.numberValue > 0) {
+    if (this->playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* obj = this->md->object(this->playerBuildings[i]);
             if (obj != 0) {
@@ -5688,7 +5512,7 @@ RGE_Static_Object* TribeInformationAIModule::findClosestReturnDropsite(RGE_Stati
             }
 
             i++;
-        } while (i < this->playerBuildings.numberValue);
+        } while (i < this->playerBuildings.number());
     }
 
     return result;
@@ -5821,7 +5645,7 @@ int TribeInformationAIModule::numberStoragePits() {
     int i = 0;
     int result = 0;
 
-    if (playerBuildings.numberValue > 0) {
+    if (playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* obj = md->object(playerBuildings[i]);
             if (obj != 0 && obj->master_obj->id == 0x67) {
@@ -5829,7 +5653,7 @@ int TribeInformationAIModule::numberStoragePits() {
             }
 
             i++;
-        } while (i < playerBuildings.numberValue);
+        } while (i < playerBuildings.number());
     }
 
     return result;
@@ -5841,7 +5665,7 @@ int TribeInformationAIModule::numberGranaries() {
     int i = 0;
     int result = 0;
 
-    if (playerBuildings.numberValue > 0) {
+    if (playerBuildings.number() > 0) {
         do {
             RGE_Static_Object* obj = md->object(playerBuildings[i]);
             if (obj != 0 && obj->master_obj->id == 0x44) {
@@ -5849,7 +5673,7 @@ int TribeInformationAIModule::numberGranaries() {
             }
 
             i++;
-        } while (i < playerBuildings.numberValue);
+        } while (i < playerBuildings.number());
     }
 
     return result;
@@ -6209,40 +6033,28 @@ int TribeInformationAIModule::resourceTypeToPlaceDropsiteBy(int param_1) {
 }
 
 int TribeInformationAIModule::dropsitesWithinRequiredDistance(
-    int& numApplicableResources, int resourceType, int requiredDistance)
+    int& numDS, int rType, int maxDistance)
 {
-    int totalApplicable;
-    int index;
-
+    int numApplicableResources;
+    numDS = 0;
     numApplicableResources = 0;
-    totalApplicable = 0;
-    ManagedArray_int dropsites;
-    index = 0;
-
-    if (this->numResources[resourceType] > 0) {
-        do {
-            if (this->resources[resourceType][index].gone != 1) {
-                if (this->md->object(this->resources[resourceType][index].id) != 0) {
-                    ++totalApplicable;
-                    if ((int)this->resources[resourceType][index].dropDistance <= requiredDistance) {
-                        if (dropsites.contains(this->resources[resourceType][index].dropsiteID) == 0) {
-                            ++numApplicableResources;
-                            int dropsiteID = this->resources[resourceType][index].dropsiteID;
-                            dropsites.add(dropsiteID);
-                        }
+    ManagedArray<int> dropsites;
+    for (int a = 0; a < this->numResources[rType]; a++) {
+        if (this->resources[rType][a].gone != 1) {
+            if (this->md->object(this->resources[rType][a].id) != 0) {
+                numApplicableResources++;
+                if ((int)this->resources[rType][a].dropDistance <= maxDistance) {
+                    if (dropsites.contains(this->resources[rType][a].dropsiteID) == 0) {
+                        numDS++;
+                        dropsites.add(this->resources[rType][a].dropsiteID);
                     }
                 }
             }
-
-            ++index;
-        } while (index < this->numResources[resourceType]);
+        }
     }
-
-    if (totalApplicable > 0) {
+    if (numApplicableResources > 0)
         return 1;
-    } else {
-        return 0;
-    }
+    return 0;
 }
 
 int TribeInformationAIModule::numberAvailableStoragePits(int param_1)
